Sources/NIOSSL/NIOSSLHandler.swift

```diff
@@ -693,6 +693,17 @@ extension NIOSSLHandler {
     private typealias BufferedWrite = (data: ByteBuffer, promise: EventLoopPromise<Void>?)
 
     private func bufferWrite(data: ByteBuffer, promise: EventLoopPromise<Void>?) {
+        var data = data
+
+        // Here we guard against the possibility that any of these writes are larger than CInt.max.
+        // This is very unusual but it can happen. To work around it, we just pretend that there were
+        // multiple writes.
+        let maxWriteSize = Int(CInt.max)
+        while let slice = data.readSlice(length: maxWriteSize) {
+            bufferedWrites.append((data: slice, promise: promise))
+        }
+
+        assert(data.readableBytes <= maxWriteSize)
         bufferedWrites.append((data: data, promise: promise))
```

Comment:
This will append an empty slice if `data.readableBytes % maxWriteSize == 0`. Should we guard against empty buffer appends?

Reply:
Oh shoot I think we're mishandling promises here too. Another patch inbound.

Reply:
To be clear, empty writes are fine here: we just need to handle promises properly.

Reply:
Oh you are right! Good catch!